### PR TITLE
SemanticMasks to Image - Diffusion Model

### PR DESCRIPTION
**Pull request by FrsECM**
# What does this PR do?

In this PR, i'll try to implement a SemanticMask to Image model like in this paper :
https://arxiv.org/abs/2207.00050
image

Implemented originaly by : 
[Weilun Wang](https://scholar.google.com/citations?hl=zh-CN&user=YfV4aCQAAAAJ), [Jianmin Bao](https://scholar.google.com/citations?hl=zh-CN&user=hjwvkYUAAAAJ), [Wengang Zhou](https://scholar.google.com/citations?hl=zh-CN&user=8s1JF8YAAAAJ), [Dongdong Chen](https://scholar.google.com/citations?hl=zh-CN&user=sYKpKqEAAAAJ), [Dong Chen](https://scholar.google.com/citations?hl=zh-CN&user=_fKSYOwAAAAJ), [Lu Yuan](https://scholar.google.com/citations?hl=zh-CN&user=k9TsUVsAAAAJ), [Houqiang Li](https://scholar.google.com/citations?hl=zh-CN&user=7sFMIKoAAAAJ),

Include # (issue)
- [x] Pipeline Semantic Only Diffusion
- [x] UNet2DSISModel
- [x] Training Script
- [x] Custom DDPM Scheduler dedicated for training

Feel free to comment and review, this PR is just a draft yet but i hope to make it work.
Thanks

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sayakpaul @patrickvonplaten

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sayakpaul** [comment]
Hi!

Thanks for willing to contribute this. Could we maybe start with a community pipeline first? If we see enough usage, we can definitely graduate it to become a core pipeline. 

Starting with a community pipeline is easier for both the contributor and us, the maintainers. 

Additionally, out of curiosity, can't we use a ControlNet model here? We did see some evidence this year that it can actually work. See here: https://huggingface.co/spaces/mfidabel/controlnet-segment-anything.

**Comment by FrsECM** [comment]
> Hi!
> 
> Thanks for willing to contribute this. Could we maybe start with a community pipeline first? If we see enough usage, we can definitely graduate it to become a core pipeline.
> 
> Starting with a community pipeline is easier for both the contributor and us, the maintainers.
> 
> Additionally, out of curiosity, can't we use a ControlNet model here? We did see some evidence this year that it can actually work. See here: https://huggingface.co/spaces/mfidabel/controlnet-segment-anything.

Hi ! 
Thanks for your support.
I'm not very fluent with the organisation of the repo. 
I can definitely move it to the community part, **just where to put it in the library ?** 
I can do this without problem : 


Just i wonder about **model** and **pipelines** : 


**Should i put it in the same folder as the rest?**

The issue i have is that i have a custom model because of the architecture in the paper : 


Concerning your question about using an existing pipeline...
The problem with other diffusion method is that they are all based on a text prompt guidance what is not interesting when you just want to do semantic inversion (in order to generate fake samples to improve a semantic segmentation model for example). 

I work in the industry and describe what we want to generate with open vocabulary is impossible because our language / wording is very specific and the language is not rich enough to retrain a text encoder. We would like to use semantic inversion in order to transfer some specific textures from one image to the other for example.

Hope it gives you a little bit of context.
Thanks !

**Comment by sayakpaul** [comment]
> Why not using existing like ControlNet ?

Makes perfect sense! Thank you for providing the details. 

Regarding your question, we have a section on contributing community pipelines here:
https://huggingface.co/docs/diffusers/conceptual/contribution#6-contribute-a-community-pipeline. In short, under the [[`community`](https://github.com/huggingface/diffusers/tree/main/examples/community)] folder, you can create another folder and put everything related to the pipeline implementation there, including a README detailing the usage. 

Does that help?

**Comment by FrsECM** [comment]
> Regarding your question, we have a section on contributing community pipelines here:

Perfect ! I'll modify in that way.

**Comment by FrsECM** [comment]
After all theses fixes, we finally successfully generates images in resolution 64x64 : 



We have tests in progress in order to increase the resolution.

**Comment by FrsECM** [comment]
@sayakpaul, it seems i have problems with black.
Do you implement an automatic linting in order to pass every tests pipelines in the CI ?

Thanks !

**Comment by sayakpaul** [comment]
@FrsECM you can run `make style && make quality` to fix the styling related failures. More info:
https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md
